### PR TITLE
Adds HOWTO "Convert PyTorch Models to Flax" to docs

### PR DESCRIPTION
**Pull request by matthias-wright**
This HOWTO shows how to convert PyTorch models to Flax.
The idea for this HOWTO originated in #1770, specifically [this](https://github.com/google/flax/discussions/1770#discussioncomment-1942875) comment by @marcvanzee.

### CODE REVIEW

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+the zero-padding will not be considered for the average calculation. There does not exist a similar
+parameter for |flax.linen.avg_pool()|_. However, we can easily implement a wrapper around the pooling
+operation.
+
+.. |flax.linen.avg_pool()| replace:: ``flax.linen.avg_pool()``
+.. _flax.linen.avg_pool(): https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.avg_pool.html
+
+.. testcode::
+
+   def avg_pool(inputs, window_shape, strides=None, padding='VALID'):
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:229]: should we also test this with an assert?

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+
+
+Batch Norm
+--------------------------------
+
+``torch.nn.BatchNorm2d`` uses ``0.1`` as the default value for the momentum parameter while
+|flax.linen.BatchNorm|_ uses ``0.9``. However, this corresponds to the same computation, because PyTorch multiplies
+the estimated statistic with ``(1 − momentum)`` and the new observed value with ``momentum``,
+while Flax multiplies the estimated statistic with momentum and the new observed value with ``(1 − momentum)``.
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:180]: nit: ``momentum``

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+  import numpy as np
+  import jax
+  from jax import random, numpy as jnp
+  import flax
+
+  from flax import linen as nn
+  from flax.core import freeze
+
+  import torch
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:14]: this is currently failing because `torch` is not listed as a `[testing]` dependency.

could you add it to to `setup.py` in the same PR?
(we'd have to look at overall check setup and run time to see if this causes a significant change before deciding if we want to add it for real)

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  kernel = t_fc.weight.detach().cpu().numpy()
+  bias = t_fc.bias.detach().cpu().numpy()
+  
+  # [outC, inC] -> [inC, outC]
+  kernel = jnp.transpose(kernel, (1, 0))
+  
+  key = random.PRNGKey(0)
+  x = random.normal(key, (1, 3))
+  
+  params = freeze({'params': {'kernel': kernel, 'bias': bias}})
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:38]: you can omit freeze here

either you should say
```
variables = {'params': {'kernel': kernel, 'bias': bias}}
```

or
```
params = {'kernel': kernel, 'bias': bias}
j_fc.apply({'params': params}, x)
```

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+  j_fc = nn.Dense(features=4)
+  j_out = j_fc.apply(params, x)
+  
+  t_x = torch.from_numpy(np.array(x))
+  t_out = t_fc(t_x)
+  t_out = t_out.detach().cpu().numpy()
+  
+  assert np.all(np.abs(j_out - t_out) < 1e-06)
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:46]: nit:

```
np.testing.assert_almost_equal(j_out, t_out) 
```

File: docs/howtos/convert_pytorch_to_flax.rst (hunk 4 above)
Comment (andsteing) [reply docs/howtos/convert_pytorch_to_flax.rst:38]: (same comment applies below as well)

File: docs/howtos/convert_pytorch_to_flax.rst (hunk 5 above)
Comment (andsteing) [reply docs/howtos/convert_pytorch_to_flax.rst:46]: (same comment applies below as well)

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+
+
+
+Batch Norm
+--------------------------------
+
+``torch.nn.BatchNorm2d`` uses ``0.1`` as the default value for the momentum parameter while
+|flax.linen.BatchNorm|_ uses ``0.9``. However, this corresponds to the same computation, because PyTorch multiplies
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:178]: nit: in the text you use `flax.linen` but in the code you use `nn` - consider using the same name in both text and code to avoid any confusion in that matter

Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:178]: nit: in the text you use `flax.linen` but in the code you use `nn` - consider using the same name in both text and code to avoid any confusion in that matter

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+     y = y / counts 
+     return y
+
+
+
+
+Transposed Convolutions
+--------------------------------
+
+``torch.nn.ConvTranspose2d`` and |flax.linen.ConvTranspose|_ are not compatible.
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:267]: This paragraph will make the reader wonder how a `torch.nn.ConvTranspose2d` should be converted. Do you know how?

**Comment by andsteing** [comment]
This is super useful! Thanks so much for writing this up...

**Review by andsteing — CHANGES_REQUESTED** [review]
Great HOWTO, added some comments...

**Comment by matthias-wright** [comment]
Thanks for the quick review!

Regarding the paragraph on ``torch.nn.ConvTranspose2d``: there is an implementation of a gradient based transposed convolution but the [PR](https://github.com/google/jax/pull/5772) is currently pending. For now, I just linked to the PR in the HOWTO. In order to create a code example, I would have to copy the implementation from the PR into the HOWTO. But the implementation consists of 207 lines of code. I think that might make the HOWTO a bit messy. What do you think?

File: docs/howtos/convert_pytorch_to_flax.rst (hunk 3 above)
Comment (andsteing) [reply docs/howtos/convert_pytorch_to_flax.rst:13]: Note: adding `torch` adds about 30 seconds to the `Install Dependencies` step in the build. The build time variance is rather large (12 min ... 20 min)

File: docs/howtos/convert_pytorch_to_flax.rst (hunk 7 above)
Comment (andsteing) [reply docs/howtos/convert_pytorch_to_flax.rst]: Thanks for adding the reference. I created #1872 to track this.

**Review by andsteing — CHANGES_REQUESTED** [review]
Thanks for the updates! One pending comment about the PyTorch compatible `avg_pool()` implementation, otherwise PR looks good to me.

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+``torch.nn.AvgPool2d`` and |nn.avg_pool()|_ are compatible when using default parameters.
+However, ``torch.nn.AvgPool2d`` has a parameter ``count_include_pad``. When ``count_include_pad=False``,
+the zero-padding will not be considered for the average calculation. There does not exist a similar
+parameter for |nn.avg_pool()|_. However, we can easily implement a wrapper around the pooling
+operation.
+
+
+.. testcode::
+
+  def pool(inputs, init, reduce_fn, window_shape, strides, padding):
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:224]: In the interest of keeping this short and easy to read, let's instead export the function as `nn.pool()`

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+.. _nn.pooling: https://github.com/google/flax/blob/main/flax/linen/pooling.py
+
+.. testcode::
+  
+  def avg_pool(inputs, window_shape, strides=None, padding='VALID'):
+    """
+    Pools the input by taking the average over a window.
+    In comparison to nn.avg_pool(), this pooling operation does not
+    consider the padded zero's for the average computation.
+    Args:
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:293]: nit: still in the interest of keeping this short and easy to read, let's just remove the `Args: ...` pydoc

File: docs/howtos/convert_pytorch_to_flax.rst [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+      window_shape: a shape tuple defining the window to reduce over.
+      strides: a sequence of `n` integers, representing the inter-window
+          strides (default: `(1, ..., 1)`).
+      padding: either the string `'SAME'`, the string `'VALID'`, or a sequence
+        of `n` `(low, high)` integer pairs that give the padding to apply before
+        and after each spatial dimension (default: `'VALID'`).
+    Returns:
+      The average for each window slice.
+    """
+    assert inputs.ndim == 4
```
Comment (andsteing) [inline docs/howtos/convert_pytorch_to_flax.rst:304]: nit: I think the updated implementation should work with arbitrary `ndim` like the `nn.avg_pool()`

**Comment by matthias-wright** [comment]
Thanks for the feedback!

**Review by andsteing — APPROVED** [review]
